scalalib/src/mill/scalalib/cosmopolitan/CosmopolitanModule.scala

```diff
@@ -0,0 +1,131 @@
+        """.stripMargin
+      }.mkString("\n");
+
+    val preArgvSize = size + 3;
+
+    finalMainClassOpt().toOption.map { cls =>
+      s"""#include <stdlib.h>
+         |#include <stdio.h>
+         |#include <errno.h>
+         |
+         |int main(int argc, char* argv[]) {
+         |  size_t preargv_size = ${preArgvSize};
+         |  size_t total = preargv_size + argc;
+         |  char *all_argv[total];
+         |  memset(all_argv, 0, sizeof(all_argv));
+         |
+         |  all_argv[0] = (char*)malloc((strlen(argv[0]) + 1) * sizeof(char));
+         |  strcpy(all_argv[0], argv[0]);
+         |  ${addForkArgs}
+         |  ${addForkArgsArgv0}
+         |  all_argv[${size}] = (char*)malloc((strlen("-cp") + 1) * sizeof(char));
+         |  strcpy(all_argv[${size}], "-cp");
+         |  all_argv[${size + 1}] = (char*)malloc((strlen(argv[0]) + 1) * sizeof(char));
+         |  strcpy(all_argv[${size + 1}], argv[0]);
+         |  all_argv[${size + 2}] = (char*)malloc((strlen("${cls}") + 1) * sizeof(char));
+         |  strcpy(all_argv[${size + 2}], "${cls}");
+         |
+         |  int i = preargv_size;
+         |  for (int count = 1; count < argc; count++) {
+         |    all_argv[i] = (char*)malloc((strlen(argv[count]) + 1) * sizeof(char));
+         |    strcpy(all_argv[i], argv[count]);
+         |    i++;
+         |  }
+         |
+         |  all_argv[total - 1] = NULL;
+         |
+         |  execvp("java", all_argv);
+         |  if (errno == ENOENT) {
+         |    execvp("java.exe", all_argv);
+         |  }
+         |
+         |  perror("java");
+         |  exit(EXIT_FAILURE);
+         |}
+      """.stripMargin
+    }
+  }
+
+  def apeCompiledLauncherScript: T[Option[PathRef]] = Task {
+    apeLauncherScript().map { sc =>
+      os.write(Task.dest / "launcher.c", sc)
+      os.call((cosmocc().path, "-mtiny", "-o", Task.dest / "launcher", Task.dest / "launcher.c"))
+
+      PathRef(Task.dest / "launcher")
+    }
+  }
+
+  def apeAssembly: T[PathRef] = Task {
+    val prepend: Option[os.Source] =
+      apeCompiledLauncherScript().map(p => os.read.inputStream(p.path))
+    val upstream = upstreamAssembly()
+
+    val created = Assembly.create0(
+      destJar = Task.dest / "out.jar.exe",
+      Seq.from(localClasspath().map(_.path)),
+      manifest(),
+      prepend,
+      Some(upstream.pathRef.path),
+      assemblyRules
+    )
+    // See https://github.com/com-lihaoyi/mill/pull/2655#issuecomment-1672468284
+    val problematicEntryCount = 65535
+    if (
+      prepend.isDefined &&
+      (upstream.addedEntries + created.addedEntries) > problematicEntryCount
+    ) {
+      Result.Failure(
+        s"""The created assembly jar contains more than ${problematicEntryCount} ZIP entries.
+           |JARs of that size are known to not work correctly with a prepended shell script.
+           |Either reduce the entries count of the assembly or disable the prepended shell script with:
```

Comment:
We should reword this, since it's mentioning of "shell script" is misleading. Best make this specific, since the user doesn't what to prepend a shell script, she wants to build a universal/portable executable.

Reply:
It's probably warranted to just fail the task in case the JAR has to many entries.

Reply:
Good catch!